src/read/magic_finder.rs

```diff
@@ -0,0 +1,191 @@
+use std::io::{Read, Seek, SeekFrom};
+
+use memchr::memmem::FinderRev;
+
+use crate::result::ZipResult;
+
+/// A utility for finding magic symbols from the end of a seekable reader.
+///
+/// Can be repurposed to recycle the internal buffer.
+pub struct MagicFinder<'a> {
+    buffer: Box<[u8]>,
+    pub(self) finder: FinderRev<'a>,
+    cursor: u64,
+    mid_buffer_offset: Option<usize>,
+    bounds: (u64, u64),
+}
+
+impl<'a> MagicFinder<'a> {
+    /// Create a new magic bytes finder to look within specific bounds.
+    pub fn new(magic_bytes: &'a [u8], bounds: (u64, u64)) -> Self {
+        const BUFFER_SIZE: usize = 2048;
+
+        // Smaller buffer size would be unable to locate bytes.
+        // Equal buffer size would stall (the window could not be moved).
+        debug_assert!(BUFFER_SIZE > magic_bytes.len());
```

Comment:
Should actually be `2 * BUFFER_SIZE - 1`, to ensure that if the entire magic couldn't fit into the window before shifting the window, it can afterward.

Reply:
On each windows pass, the cursor is moved by `BUFFER_SIZE - magic_bytes.len()` back, so it will contain magic bytes at the boundary. Looking at it now, we actually must move the window by `BUFFER_SIZE - magic_bytes.len() + 1` to not count magic bytes exactly at the start of the window twice. The actuall assertion for the window to move should then be `BUFFER_SIZE >= magic_bytes.len()`.